spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/bigquant/SpatialConvolution.scala

```diff
@@ -0,0 +1,382 @@
+
+    val weightFP32Tmp = weightFP32.view(Array(nOutputPlane, nInputPlane / nGroup, kernelH, kernelW))
+    for (i <- 1 to nOutputPlane) {
+      val singleRow = weightFP32Tmp.select(1, i)
+      min(i - 1) = singleRow.min()
+      max(i - 1) = singleRow.max()
+    }
+
+    weight = new Array[QuantTensor[T]](nGroup)
+    for (i <- 0 until nGroup) {
+      weight(i) = new QuantTensor[T](nOutputPlane / nGroup, nInputPlane / nGroup, kernelH, kernelW)
+    }
+
+    for (i <- 1 to nGroup) {
+      val groupWeight = weightFP32.select(1, i)
+      val bufferOffset = 0
+      val buffer = new Array[Byte](groupWeight.nElement())
+      ev.getType() match {
+        case FloatType =>
+          Quant.quantize(groupWeight.toTensor[Float], buffer, bufferOffset)
+        case _ => throw new UnsupportedOperationException(s"Only support Float for quantized model")
+      }
+
+      weight(i - 1).setStorage(buffer)
+    }
+
+    init()
+
+    this
+  }
+
+  @throws(classOf[IOException])
+  private def writeObject(out: ObjectOutputStream): Unit = {
+    out.defaultWriteObject()
+
+    out.writeObject(weight)
+  }
+
+  @throws(classOf[IOException])
+  private def readObject(in: ObjectInputStream): Unit = {
+    in.defaultReadObject()
+
+    weight = in.readObject().asInstanceOf[Array[QuantTensor[T]]]
+    if (weight(0).getStorage != null && weight(0).getStorageInJni == 0L) {
+      init()
+    }
+  }
+
+  private def init(): this.type = {
+    for (i <- 1 to nGroup) {
+      if (weight(i - 1).getStorageInJni == 0L) {
+        val byteArrayOfWeight = weight(i - 1).getStorage
+        weight(i - 1).setStorageInJni(
+          BigQuant.ConvKernelDescInit(nOutputPlane / nGroup, nInputPlane / nGroup,
+            kernelH, kernelW))
+        ev.getType() match {
+          case FloatType =>
+            val start = (i - 1) * nOutputPlane / nGroup
+            val end = i * nOutputPlane / nGroup
+            val minArray = min.asInstanceOf[Array[Float]].slice(start, end)
+            val maxArray = max.asInstanceOf[Array[Float]].slice(start, end)
+            val byteOffset = 0
+
+            BigQuant.ConvKernelLoadFromModel(weight(i - 1).getStorageInJni,
+              byteArrayOfWeight, byteOffset,
+              minArray, maxArray, nOutputPlane / nGroup, nInputPlane / nGroup,
+              kernelH, kernelW, WEIGHT_THRESHOLD, BigQuant.NCHW)
+          case _ =>
+            throw new UnsupportedOperationException(s"Only support Float for quantized model")
+        }
+      }
+    }
+
+    _init = true
+
+    this
+  }
+
+  private def checkAndInit(): Unit = {
+    if (!_init) {
+      init()
+    }
+  }
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    require(input.dim() == 3 || input.dim() == 4,
+      "bigquant.SpatialConvolution: " + ErrorInfo.constrainInputAs3DOrBatch)
+    require(input.isContiguous())
+
+    checkAndInit()
+
+    // compute attributes of input and output
+    val (batchSize, inputHeight, inputWidth) = if (input.dim() == 3) {
```

Comment:
Only NCWH is supported, yes?

Reply:
NHWC also be supported. This feature will be added soon.